In [1]:
# import/load packages

## install
%pip install pandas
%pip install chardet
%pip install cchardet


## import
import os
import pandas as pd
import chardet
import cchardet
import re
from bs4 import BeautifulSoup


Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [3]:
# see directory
os. getcwd()

'/work/Bachelor'

## Check that dataframe structures are **proper**

## Testing 05-17 text.txt and sources.txt files from sample data

### sources-17-05.txt

In [4]:
# sources

# define file path
file_path = "/work/Bachelor/sample_data/dropbox/sources-17-05.txt"

# read csv file into dataframe
df_sources_17_05 = pd.read_csv(file_path, delimiter='\t', encoding='ISO-8859-1')

# define column names
column_names = ["textID", "words", "date", "country", "source", "url", "title"]

# assign new column names to the dataframe
df_sources_17_05.columns = column_names
df_sources_17_05

textID  words      date country               source  \
0       18207168   2362  17-05-01      US      Huffington Post   
1       18207169    343  17-05-01      US          GoLocalProv   
2       18207170    257  17-05-01      US             Engadget   
3       18207172    180  17-05-01      US     Northwest Herald   
4       18207174    630  17-05-01      US          GoLocalProv   
...          ...    ...       ...     ...                  ...   
297859  18740662    304  17-05-31      JM  Jamaica Star Online   
297860  18742209    315  17-05-31      JM      Jamaica Gleaner   
297861  18742210    127  17-05-31      JM  Jamaica Star Online   
297862  18742211    288  17-05-31      JM  Jamaica Star Online   
297863  18741456    175  17-05-31      JM      Jamaica Gleaner   

                                                      url  \
0       http://www.huffingtonpost.com/entry/a-letter-t...   
1       http://www.golocalprov.com/business/chafee-ste...   
2       https://www.engadget.com/2017/05/01/lab-grown-...   
3       http://www.nwherald.com/2017/04/28/friends-of-...   
4       http://www.golocalprov.com/politics/guest-mind...   
...                                                   ...   
297859  http://jamaica-star.com/article/news/20170531/...   
297860  http://jamaica-gleaner.com/article/news/201705...   
297861  http://jamaica-star.com/article/entertainment/...   
297862  http://jamaica-star.com/article/entertainment/...   
297863  http://jamaica-gleaner.com/article/entertainme...   

                                                    title  
0                   A letter to parents of expat children  
1       Chafee Steps Up Criticism of Raimondo's Econom...  
2       Lab-grown stem cells may carry an increased ri...  
3       Friends of McHenry County College Foundation t...  
4       Guest MINDSETTERT James Monteiro: Vote No in R...  
...                                                   ...  
297859  Christopher Townsend is back on Munga's murder...  
297860  OUR proposes fixes to JPS after August 2016 bl...  
297861           Journalist curses Damian Marley in Kenya  
297862  DON'T PRAY FOR ME ... Pepita slams unworthy Ch...  
297863  Reggae could be added to UNESCO's list of Inta...  

[297864 rows x 7 columns]

In [3]:
# check how many articles there are from each country
country_counts = df_sources_17_05['country'].value_counts()
print(country_counts)

country
IN    31756
CA    30205
US    26663
GB    26223
IE    25204
NG    21151
AU    20726
NZ    18999
ZA    18324
MY    17759
SG    16959
PH    14018
PK    10431
KE     7038
GH     6066
HK     1853
LK     1762
JM     1352
BD     1279
TZ       96
Name: count, dtype: int64


In [5]:
# filter to only include sources from the US
dropbox_sources_17_05_US = df_sources_17_05[df_sources_17_05['country'] == 'US']

# check length to compare with the dropbox_text file
len(dropbox_sources_17_05_US)


26663

In [3]:
# save as csv to check that articles aren't nested in each other

# Define the file path where you want to save the CSV
output_file_path = "/work/Bachelor/sample_data/stopthenonsense/dropbox_sources_17_05_US.csv"

# Save the filtered DataFrame as a CSV file
dropbox_sources_17_05_US.to_csv(output_file_path, index=False, encoding='utf-8')

NameError: name 'dropbox_sources_17_05_US' is not defined

### text_17-05-US.txt

In [23]:
# not
# text

# Define file path
file_path = "/work/Bachelor/sample_data/dropbox/text_17-05-US.txt"

# Open the file and read its contents into a string
with open(file_path, 'r', encoding='utf-8') as file:
    raw_text = file.read()  # Read all text data from the file

# Step 1: Split the text into articles based on '@@' markers
articles = raw_text.split('@@')[1:]  # Split the text by '@@' and ignore the first empty item

# Lists to store extracted data
text_ids = []  # Initialize a list to store article IDs
clean_texts = []  # Initialize a list to store cleaned article texts

# Step 2: Process each article
for article in articles:  # Loop through each article in the list
    # Extract article ID (the number right after @@)
    article_id_match = re.match(r'(\d+)', article)  # Match numbers at the beginning of the article
    text_id = article_id_match.group(1) if article_id_match else None  # Extract ID if match found, else None

    # Remove the ID and any leading characters (up to the first space)
    cleaned_article = re.sub(r'^\d+\s*', '', article)  # Remove leading numbers and spaces

    # Remove HTML-like tags (<h>, <p>, etc.)
    cleaned_article = re.sub(r'<[^>]+>', '', cleaned_article).strip()  # Remove HTML tags and extra spaces

    # Append extracted data to lists
    text_ids.append(text_id)  # Add the ID to the text_ids list
    clean_texts.append(cleaned_article)  # Add the cleaned text to the clean_texts list

# Step 3: Create a DataFrame to store the structured data
dropbox_text_17_05 = pd.DataFrame({
    'textID': text_ids,  # Column with article IDs
    'text': clean_texts   # Column with cleaned article texts
})

# Display the DataFrame
dropbox_text_17_05  # Output the DataFrame to view structured data


textID                                               text
0      18207159  It appears that right now is a pretty good tim...
1      18207168  A letter to parents of expat children  Parenti...
2      18207169  Chafee Steps Up Criticism of Raimondo 's Econo...
3      18207170  Lab-grown stem cells may carry an increased ri...
4      18207172  Friends of McHenry County College Foundation t...
...         ...                                                ...
27082  18745486  1954 : American film star Marilyn Monroe ( 192...
27083  18745488  On the two-year anniversary of his murder , Ka...
27084  18745489  Artists file lawsuit against city of Atlanta o...
27085  18745490  MONTESANO , Wash . -- James Walker , 31 , of H...
27086  18745491  Out to the track for our KSFY/Taco John 's Ath...

[27087 rows x 2 columns]

In [47]:
# define file path
file_path = "/work/Bachelor/sample_data/dropbox/text_17-05-US.txt"

# open the file and read its contents into a string
with open(file_path, 'r', encoding='utf-8') as file:
    raw_text = file.read()  # Read all text data from the file

sample = raw_text
sample = re.sub(" ([.,?!':])", r"\1", sample) # punctuation characters are all needlessly preceded by a space, remove that space
sample = re.sub("@ @ @ @ @ @ @ @ @ @", "CENSORED", sample) # replace the keyword chosen to mark words or entities that have been censored

# Step 1: Split the text into articles based on '@@' markers
article_ids = re.findall(r"@@(\d+)", sample) # extract list of all article IDs by matching pattern of any sequence of digits following exactly two @
articles = re.split(r'"?@@\d+ ', sample)[1:] # split articles on article IDs, i.e. a sequence of digits preceded by exactly two @s and sometimes a quotation mark preceding that
articles = [art[art.find("<p> ") + 4:].strip().replace(" <p> ", "\n") for art in articles] # for each article, skip to the first paragraph tag to exclude heading/title text, then replace subsequent markers for new paragraphs with line breaks 

print(len(article_ids), len(articles)) # check that we found equally many IDs and article bodies

articles_US = pd.DataFrame(data = dict(textID = article_ids, body = articles)) # create dataframe with extracted IDs and corresponding article contents
articles_US["textID"] = articles_US["textID"].astype(int) # cast textIDs to int to enable merging with `sources` dataframe
articles_US.sample(10) # sample ten random rows from the dataframe to examine whether things look as expected


27087 27087
18207159
It appears that right now is a pretty good time to be a band called Dawes. With the release of their

18207168
Parenting is fraught with anxiety, joys, challenges.... Parenting expat children is a different ball

18207169
In a letter sent to members of the Board of the Commerce Corporation, former Governor Lincoln Chafee

18207170
If you've followed the latest medical research, you know that stem cells are a big deal. They let yo

18207172
The Friends of McHenry County College Foundation is hosting its 2017 Friends of MCC Foundation Golf 

18207174
I write to go on the record and say that I've been a Ward 3 resident probably longer than at least 9

18207178
Connor Mitchell, right, works on a computer at a cafe in San Francisco on April 13, 2017. With colle

18207182
Personal hero and why: My dad, because he is a volunteer firefighter and gives his free time to help

18207185
Students named in the complaints that disqualified former Student Association presidential 

textID                                               body
8261   18363418  TWIN FALLS -- Businesses that have a skills ga...
1373   18232116  Approximately one hundred Internet years ago -...
22100  18643620  CLEVELAND -- As the nation's deadly opioid cri...
2363   18251282  The old saying in regards to Abraham Lincoln's...
20903  18619016  A secret document that officials say played a ...
24461  18687854  Traffic backups on Oracle Road, like this one ...
6884   18334023  Ahead of a site visit by the International Oly...
6544   18326663  The law has to go through further drafting by ...
1357   18231485  "" It was awful to watch that.... People shoul...
26601  18736927  Updated: May 31, 2017 - 1:14 PM\nPITTSBURGH - ...

In [48]:
joined = dropbox_sources_17_05_US.merge(articles_US, on = "textID", how = "left") # join sources dataframe with their corresponding article content using the textIDs, keeping only the articles where source information was available
joined

textID  words      date country            source  \
0      18207168   2362  17-05-01      US   Huffington Post   
1      18207169    343  17-05-01      US       GoLocalProv   
2      18207170    257  17-05-01      US          Engadget   
3      18207172    180  17-05-01      US  Northwest Herald   
4      18207174    630  17-05-01      US       GoLocalProv   
...         ...    ...       ...     ...               ...   
26658  18745486   1212  17-05-31      US        OCRegister   
26659  18745488    424  17-05-31      US        fox4kc.com   
26660  18745489    405  17-05-31      US     FOX 5 Atlanta   
26661  18745490    458  17-05-31      US           Q13 FOX   
26662  18745491    150  17-05-31      US              KSFY   

                                                     url  \
0      http://www.huffingtonpost.com/entry/a-letter-t...   
1      http://www.golocalprov.com/business/chafee-ste...   
2      https://www.engadget.com/2017/05/01/lab-grown-...   
3      http://www.nwherald.com/2017/04/28/friends-of-...   
4      http://www.golocalprov.com/politics/guest-mind...   
...                                                  ...   
26658  http://www.ocregister.com/2017/05/31/rememberi...   
26659  http://fox4kc.com/2017/05/31/on-the-two-year-a...   
26660    http://www.fox5atlanta.com/news/258189601-story   
26661  http://q13fox.com/2017/05/31/man-accused-of-ba...   
26662  http://www.ksfy.com/content/sports/Brandon-Val...   

                                                   title  \
0                  A letter to parents of expat children   
1      Chafee Steps Up Criticism of Raimondo's Econom...   
2      Lab-grown stem cells may carry an increased ri...   
3      Friends of McHenry County College Foundation t...   
4      Guest MINDSETTERT James Monteiro: Vote No in R...   
...                                                  ...   
26658  Remembering Marilyn Monroe on anniversary of h...   
26659  On the two-year anniversary of his murder, Kan...   
26660  Artists file lawsuit against city of Atlanta o...   
26661  Man accused of backing truck over 2 men, killi...   
26662  Brandon Valley senior Jakob Hanna named KSFY/T...   

                                                    body  
0      Parenting is fraught with anxiety, joys, chall...  
1      In a letter sent to members of the Board of th...  
2      If you've followed the latest medical research...  
3      The Friends of McHenry County College Foundati...  
4      I write to go on the record and say that I've ...  
...                                                  ...  
26658  1954: American film star Marilyn Monroe ( 1926...  
26659  RAYTOWN, Mo. -- A 3-year-old's siblings and mo...  
26660  ATLANTA - A group of artists have drawn a lega...  
26661  MONTESANO, Wash. -- James Walker, 31, of Hoqui...  
26662  Out to the track for our KSFY/Taco John's Athl...  

[26663 rows x 8 columns]

In [7]:
# not
# ensure both 'textID' columns have the same data type (convert to string)
dropbox_sources_17_05_US.loc[:, 'textID'] = dropbox_sources_17_05_US['textID'].astype(str)
dropbox_text_17_05.loc[:, 'textID'] = dropbox_text_17_05['textID'].astype(str)

# merge the two DataFrames on 'textID'
combined_dropbox_articles = pd.merge(dropbox_sources_17_05_US, dropbox_text_17_05, on='textID', how='inner')

# display
print(combined_dropbox_articles)
print(len(combined_dropbox_articles))

         textID  X.words      date country            source  \
0      18207168     2362  17-05-01      US   Huffington Post   
1      18207169      343  17-05-01      US       GoLocalProv   
2      18207170      257  17-05-01      US          Engadget   
3      18207172      180  17-05-01      US  Northwest Herald   
4      18207174      630  17-05-01      US       GoLocalProv   
...         ...      ...       ...     ...               ...   
26658  18745486     1212  17-05-31      US        OCRegister   
26659  18745488      424  17-05-31      US        fox4kc.com   
26660  18745489      405  17-05-31      US     FOX 5 Atlanta   
26661  18745490      458  17-05-31      US           Q13 FOX   
26662  18745491      150  17-05-31      US              KSFY   

                                                     url  \
0      http://www.huffingtonpost.com/entry/a-letter-t...   
1      http://www.golocalprov.com/business/chafee-ste...   
2      https://www.engadget.com/2017/05/01/lab-grow

### Stuff is happening!!! 26,663 sources, and 27,087 texts.
### Okay also cool, there are as many rows in the combined dataframe as there are in the source dataframe. So the text dataframe must be split correctly. Maybe something's up with the sources then...👩‍🍳 check out tomorrow, and also test these things on the sample data

In [ ]:
# playground 2 - make csv before filtering

# sources
file_path = "/work/Bachelor/sample_data/dropbox/sources-17-05.txt"

#df_sources_17_05 = pd.read_csv(file_path, delimiter='\t', encoding='ISO-8859-1')
#print(df_sources_17_05)

# Define the file path (you would adjust this to your actual file path)
file_path = "/work/Bachelor/sample_data/dropbox/text_17-05-US.txt"

# Open the file and read its contents into a string
with open(file_path, 'r', encoding='utf-8') as file:
    raw_text = file.read()

# Step 1: Split the text into articles based on '@@' markers
articles = raw_text.split('@@')[1:]  # Skip the first empty split

# Lists to store extracted data
text_ids = []
clean_texts = []

# Step 2: Process each article
for article in articles:
    # Extract article ID (the number right after @@)
    article_id_match = re.match(r'(\d+)', article)
    text_id = article_id_match.group(1) if article_id_match else None

    # Remove the ID and any leading characters (up to the first space)
    cleaned_article = re.sub(r'^\d+\s*', '', article)

    # Remove HTML-like tags (<h>, <p>, etc.)
    cleaned_article = re.sub(r'<[^>]+>', '', cleaned_article).strip()

    # Append extracted data to lists
    text_ids.append(text_id)
    clean_texts.append(cleaned_article)

# Step 3: Create a DataFrame to store the structured data
df_text_17_05 = pd.DataFrame({
    'textID': text_ids,
    'text': clean_texts
})

# Display the DataFrame
print(df_text_17_05)


In [72]:
# save as csv to check that articles aren't nested in each other

# Define the file path where you want to save the CSV
output_file_path = "/work/Bachelor/sample_data/stopthenonsense/dropbox_text_17_05.csv"

# Save the filtered DataFrame as a CSV file
df_text_17_05.to_csv(output_file_path, index=False, encoding='utf-8')

### I think it's good but fuck knows

## Testing original text.txt and sources.txt files from sample data👩‍🍳

In [10]:
# load the sources.csv that was made from excel and r
file_path = "/work/Bachelor/sample_data/stopthenonsense/sources.csv"
df_sources_fr = pd.read_csv(file_path)

print(df_sources_fr.head())  # Display the first few rows of the loaded DataFrame
len(df_sources_fr)

   textID  X.words        date country               source  \
0   11241      397  13.01.2006      US               Kotaku   
1   11242      757  13.01.2006      US       Michigan Radio   
2   11243      755  13.01.2006      US  New York Daily News   
3   11244     1677  13.01.2006      US       OregonLive.com   
4   21242      794  13.01.2011      US         Ars Technica   

                                                 url  \
0  http://kotaku.com/5973495/author-of-the-warrio...   
1  http://michiganradio.org/post/thats-what-they-...   
2  http://www.nydailynews.com/life-style/eats/bes...   
3  http://www.oregonlive.com/performance/index.ss...   
4  http://arstechnica.com/gadgets/2013/01/ask-ars...   

                                               title  
0  Author of The Warriors, Cult Film Adapted to H...  
1  That's What They Say: Dialect Society chooses ...  
2                        Best of New York: Croissant  
3  Reflecting on a quarter-century of growth in P...  
4  Ask Ar

2960

### I think this 👆 is the proper sources_csv file<33

In [48]:
# text

# Define the file path (you would adjust this to your actual file path)
file_path = "/work/Bachelor/sample_data/text.txt"

# Open the file and read its contents into a string
with open(file_path, 'r', encoding='utf-8') as file:
    raw_text = file.read()

# Step 1: Split the text into articles based on '@@' markers
articles = raw_text.split('@@')[1:]  # Skip the first empty split

# Lists to store extracted data
text_ids = []
clean_texts = []

# Step 2: Process each article
for article in articles:
    # Extract article ID (the number right after @@)
    article_id_match = re.match(r'(\d+)', article)
    text_id = article_id_match.group(1) if article_id_match else None
    

    # Remove the ID and any leading characters (up to the first space)
    cleaned_article = re.sub(r'^\d+\s*', '', article)

    # Remove HTML-like tags (<h>, <p>, etc.)
    cleaned_article = re.sub(r'<[^>]+>', '', cleaned_article).strip()

    # Append extracted data to lists
    text_ids.append(text_id)
    clean_texts.append(cleaned_article)

# Step 3: Create a DataFrame to store the structured data
df_text = pd.DataFrame({
    'textID': text_ids,
    'text': clean_texts
})


# Display the DataFrame
print(df_text)



        textID                                               text
0        11241  Sol Yurick , the writer whose 1965 novel " The...
1        11242  That 's What They Say : Dialect Society choose...
2        11243  A sublime croissant at French Tart in Grant Ci...
3        11244  Reflecting on a quarter-century of growth in P...
4        21242  Ask Ars : Does Facebook auto-delete content af...
...        ...                                                ...
2909  15281244  Did you know that Prime Minister Narendra Modi...
2910  15291240  Council voted Tuesday to support a new $30 per...
2911  15291242  News Local  $5M funding boost for six projects...
2912  15291243  A closed-door council meeting that reversed th...
2913  15291244  ?  Burrows helps Canucks beat Rangers 5-3 to e...

[2914 rows x 2 columns]


In [8]:
# playground 0

# text

# Define file path
file_path = "/work/Bachelor/sample_data/text.txt"

# Open the file and read its contents into a string
with open(file_path, 'r', encoding='utf-8') as file:
    raw_text = file.read()  # Read all text data from the file

# Step 1: Split the text into articles based on '@@' markers
articles = raw_text.split('@@')[1:]  # Split the text by '@@' and ignore the first empty item

# Lists to store extracted data
text_ids = []  # Initialize a list to store article IDs
clean_texts = []  # Initialize a list to store cleaned article texts

# Step 2: Process each article
for article in articles:  # Loop through each article in the list
    # Extract article ID (the number right after @@)
    article_id_match = re.match(r'(\d+)', article)  # Match numbers at the beginning of the article
    text_id = article_id_match.group(1) if article_id_match else None  # Extract ID if match found, else None

    # Remove the ID and any leading characters (up to the first space)
    cleaned_article = re.sub(r'^\d+\s*', '', article)  # Remove leading numbers and spaces

    # Remove HTML-like tags (<h>, <p>, etc.)
    cleaned_article = re.sub(r'<[^>]+>', '', cleaned_article).strip()  # Remove HTML tags and extra spaces

    # Append extracted data to lists
    text_ids.append(text_id)  # Add the ID to the text_ids list
    clean_texts.append(cleaned_article)  # Add the cleaned text to the clean_texts list

# Step 3: Create a DataFrame to store the structured data
sample_text = pd.DataFrame({
    'textID': text_ids,  # Column with article IDs
    'text': clean_texts   # Column with cleaned article texts
})

# Display the DataFrame
print(sample_text)  # Output the DataFrame to view structured data


        textID                                               text
0        11241  Sol Yurick , the writer whose 1965 novel " The...
1        11242  That 's What They Say : Dialect Society choose...
2        11243  A sublime croissant at French Tart in Grant Ci...
3        11244  Reflecting on a quarter-century of growth in P...
4        21242  Ask Ars : Does Facebook auto-delete content af...
...        ...                                                ...
2909  15281244  Did you know that Prime Minister Narendra Modi...
2910  15291240  Council voted Tuesday to support a new $30 per...
2911  15291242  News Local  $5M funding boost for six projects...
2912  15291243  A closed-door council meeting that reversed th...
2913  15291244  ?  Burrows helps Canucks beat Rangers 5-3 to e...

[2914 rows x 2 columns]


In [12]:
# playground 0 merge

# ensure both 'textID' columns have the same data type (convert to string)
df_sources_fr.loc[:, 'textID'] = dropbox_sources_17_05_US['textID'].astype(str)
sample_text.loc[:, 'textID'] = dropbox_text_17_05['textID'].astype(str)

# merge the two DataFrames on 'textID'
combined_articles = pd.merge(df_sources_fr, sample_text, on='textID', how='inner')

# display
print(combined_articles)
print(len(combined_articles))

        textID  X.words        date country               source  \
0     18207168      397  13.01.2006      US               Kotaku   
1     18207169      757  13.01.2006      US       Michigan Radio   
2     18207170      755  13.01.2006      US  New York Daily News   
3     18207172     1677  13.01.2006      US       OregonLive.com   
4     18207174      794  13.01.2011      US         Ars Technica   
...        ...      ...         ...     ...                  ...   
2908  18262873      743  16.11.2019      GB             Redbrick   
2909  18262874      301  16.11.2019      NG          360Nobs.com   
2910  18262875      203  16.11.2019      NG             Goal.com   
2911  18262876      270  16.11.2019      NG    Gistmaster (blog)   
2912  18262878      305  16.11.2020      GH             GhanaWeb   

                                                    url  \
0     http://kotaku.com/5973495/author-of-the-warrio...   
1     http://michiganradio.org/post/thats-what-they-...   
2     

In [80]:
# playground 1

# Define the file path (you would adjust this to your actual file path)
file_path = "/work/Bachelor/sample_data/text.txt"

# Open the file and read its contents into a string
with open(file_path, 'r', encoding='utf-8') as file:
    raw_text = file.read()

# Step 1: Split the text into articles based on '@@' markers
articles = raw_text.split('@@')[1:]  # Skip the first empty split
print(f"Step 1: Number of articles after splitting: {len(articles)}")

# Lists to store extracted data
text_ids = []
clean_texts = []

# Step 2: Process each article
for article in articles:
    # Extract article ID (the number right after @@)
    article_id_match = re.match(r'(\d+)', article)
    text_id = article_id_match.group(1) if article_id_match else None
    
    # Remove the ID and any leading characters (up to the first space)
    cleaned_article = re.sub(r'^\d+\s*', '', article)

    # Remove HTML-like tags (<h>, <p>, etc.)
    cleaned_article = re.sub(r'<[^>]+>', '', cleaned_article).strip()

    # Append extracted data to lists
    text_ids.append(text_id)
    clean_texts.append(cleaned_article)

    # Print progress for each article processed
    #print(f"Processed article ID: {text_id}, Current number of articles: {len(text_ids)}")

# Step 3: Create a DataFrame to store the structured data
df_text = pd.DataFrame({
    'textID': text_ids,
    'text': clean_texts
})

# Step 4: Print the final number of rows in the DataFrame
print(f"Step 4: Number of rows in the DataFrame: {df_text.shape[0]}")

# Display the DataFrame (optional)
print(df_text)


Step 1: Number of articles after splitting: 2914
Step 4: Number of rows in the DataFrame: 2914
        textID                                               text
0        11241  Sol Yurick , the writer whose 1965 novel " The...
1        11242  That 's What They Say : Dialect Society choose...
2        11243  A sublime croissant at French Tart in Grant Ci...
3        11244  Reflecting on a quarter-century of growth in P...
4        21242  Ask Ars : Does Facebook auto-delete content af...
...        ...                                                ...
2909  15281244  Did you know that Prime Minister Narendra Modi...
2910  15291240  Council voted Tuesday to support a new $30 per...
2911  15291242  News Local  $5M funding boost for six projects...
2912  15291243  A closed-door council meeting that reversed th...
2913  15291244  ?  Burrows helps Canucks beat Rangers 5-3 to e...

[2914 rows x 2 columns]


In [82]:
# playground 2

# Define the file path (you would adjust this to your actual file path)
file_path = "/work/Bachelor/sample_data/text.txt"

# Open the file and read its contents into a string
with open(file_path, 'r', encoding='utf-8') as file:
    raw_text = file.read()

# Step 1: Split the text using the pattern '@@' followed by digits
# The pattern '@@\d+' matches the '@@' followed by one or more digits
articles = re.split(r'@@\d+', raw_text)[1:]  # Skip the first empty split
print(f"Step 1: Number of articles after splitting: {len(articles)}")

# Lists to store extracted data
text_ids = []
clean_texts = []

# Step 2: Process each article
# Find all occurrences of '@@' followed by digits to extract IDs
article_ids = re.findall(r'@@(\d+)', raw_text)

for i, article in enumerate(articles):
    # Extract article ID from the pre-extracted article_ids list
    text_id = article_ids[i] if i < len(article_ids) else None
    
    # Remove HTML-like tags (<h>, <p>, etc.)
    cleaned_article = re.sub(r'<[^>]+>', '', article).strip()

    # Append extracted data to lists
    text_ids.append(text_id)
    clean_texts.append(cleaned_article)

    # Print progress for each article processed
    #print(f"Processed article ID: {text_id}, Current number of articles: {len(text_ids)}")

# Step 3: Create a DataFrame to store the structured data
df_text = pd.DataFrame({
    'textID': text_ids,
    'text': clean_texts
})

# Step 4: Print the final number of rows in the DataFrame
print(f"Step 4: Number of rows in the DataFrame: {df_text.shape[0]}")

# Display the DataFrame (optional)
print(df_text)


Step 1: Number of articles after splitting: 2914
Step 4: Number of rows in the DataFrame: 2914
        textID                                               text
0        11241  Sol Yurick , the writer whose 1965 novel " The...
1        11242  That 's What They Say : Dialect Society choose...
2        11243  A sublime croissant at French Tart in Grant Ci...
3        11244  Reflecting on a quarter-century of growth in P...
4        21242  Ask Ars : Does Facebook auto-delete content af...
...        ...                                                ...
2909  15281244  Did you know that Prime Minister Narendra Modi...
2910  15291240  Council voted Tuesday to support a new $30 per...
2911  15291242  News Local  $5M funding boost for six projects...
2912  15291243  A closed-door council meeting that reversed th...
2913  15291244  ?  Burrows helps Canucks beat Rangers 5-3 to e...

[2914 rows x 2 columns]


### compare rows

In [75]:
# source
print(df_sources_fr.iloc[1000])

# text
print(df_text.iloc[1000])

textID                                               2911244
X.words                                                  712
date                                              14.11.2008
country                                                   CA
source                                         Globalnews.ca
url        http://globalnews.ca/news/1661449/thousands-of...
title      Thousands of Israeli Arabs protest fatal polic...
Name: 1000, dtype: object
textID                                              2911244
text      What is this ?  WATCH ABOVE : Thousands of pro...
Name: 1000, dtype: object


#### Mmmmm it fits kind of, but it stops the further into the dataset we get

### Save df_text as csv to compare to sources.csv visually

In [45]:
# save text df as csv to compare with sources.csv

# Define the file path where you want to save the CSV
output_file_path = "/work/Bachelor/sample_data/stopthenonsense/text1.csv"

# Save the filtered DataFrame as a CSV file
df_text.to_csv(output_file_path, index=False, encoding='utf-8')